src/ast/nodes/MemberExpression.ts

```diff
@@ -64,19 +64,20 @@ export default class MemberExpression extends NodeBase {
 	computed: boolean;
 
 	propertyKey: ObjectPathKey;
-	variable: Variable;
-	private isBound: boolean;
-	private replacement: string;
+	variable: Variable = null;
 	private arePropertyReadSideEffectsChecked: boolean;
+	private bound: boolean;
+	private replacement: string | null;
 
 	bind() {
-		this.isBound = true;
+		if (this.bound) return;
```

Comment:
The problem is that if another node calls `reassignPath` or `forEachReturnExpressionAtPath`, it is possible that those methods are called on this node before it has been bound which means if this represents a namespace, the variable will not yet be assigned. This is hard to control as both methods are forwarded through variable assignments.

Another reason I would leave this in place is that, if time permits, one of my next goals is to inline binding into the actual tree-shaking so that
* unused nodes will not be bound, and
* in case unused code mutates included variables, these mutations will no longer taint the tree-shaking result and, at least in some scenarios, more code can be removed.

My expectation is that this will provide another slight performance boost.

Btw `Identifier`s follow a similar logic.

Reply:
Thanks, it helps a lot to get a better idea of how this logic works. I guess it could be avoided by a multi-pass process, but this does sound more optimized.